### DISCUSSION TITLE
Spline producer url unreachable from spark

### DISCUSSION THREAD

**Discussion by learnerr101** [AbsaOSS/spline#1355] [Q&A]
I used the TLDR configuration and packages given in https://absaoss.github.io/spline/. za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.12:0.5.2 and 'spark.spline.producer.url' as 'http://localhost:9090/producer'

But I get this error in my spark job:

![Screenshot 2024-07-24 003938](https://github.com/user-attachments/assets/dd62fb5f-1f07-457b-b186-3d9338429650)

Configurations:
Spark -> 2.4.2
Scala -> 2.12
Commands used:
spark-submit --packages za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.12:0.5.2 --conf "spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener" --conf "spark.spline.producer.url=http://localhost:9090/producer" AB_2.py

I have also tried spark 3.0.0 with compatible artifacts and Java version but I'm getting the same error.

**Comment by cerveada** [comment]
Try to follow steps from this guide: https://github.com/AbsaOSS/spline/discussions/1225

**Comment by learnerr101** [comment]
So these were the steps I performed:

**_Troubleshooting Spline Agent:_**
I was able to see the logs pertaining to the lineage.

**_Troubleshooting Arango DB:_**
![image](https://github.com/user-attachments/assets/040eec15-84ad-4ce9-81ad-9370951a4cce)

**_Troubleshooting Spline Server:_**
![image](https://github.com/user-attachments/assets/ae2f8ccd-7f74-4b39-962f-0fdd30078aa2)

**_Troubleshooting Spline UI:_**
![image](https://github.com/user-attachments/assets/89e22887-357f-420b-8564-37f04d82d688)


Still I'm getting the same error for some reason.

**Comment by cerveada** [comment]
In your config you use port `9090` but on the picture from server i see port `8080`

**Comment by learnerr101** [comment]
Yeah, I used commands with both the configs.

With 9090:
spark-submit --packages za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.12:0.5.2 --conf "spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener" --conf "spark.spline.producer.url=http://localhost:9090/producer" AB_2.py

With 8080:
spark-submit --packages za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.12:0.5.2 --conf "spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener" --conf "spark.spline.producer.url=http://localhost:8080/producer" AB_2.py


In both the scenarios I get the same error.

**Comment by cerveada** [comment]
In the log I see output from `LoggingLineageDispatcher` so that is working. Now switch to http dispatcher.

- You should use the url under **Producer API - Base URL** that you see in server page.
- I would use the latest release version of the agent: 2.1.0
- make sure the connection from agent to server is possible and nothing is blocking it

**Comment by learnerr101** [comment]
Yeah I tried the following command:
```
_spark-submit  
--packages za.co.absa.spline.agent.spark:spark-3.0-spline-agent-bundle_2.12:2.1.0  
--conf spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener 
--conf spark.spline.lineageDispatcher=http  
--conf spark.spline.lineageDispatcher.http.producer.url=http://localhost:8080/producer  
AB_2.py_
```

So the config as of now is:
**_Spark 3.0.0
Scala 2.12
Spline : 2.1.0_**

But error still prevails!

**Comment by cerveada** [comment]
If everything else works, it must be some networking issue. Are you sure that the server is reachable? Is the spark running locally on the same machine as the severe? If spark is somewhere else, localhost will not work.

Also, if a server or agent is running in the docker container, it may block the connection.

**Reply by learnerr101** [reply]
Yeah I can access:
**_http://localhost:9090/    -> UI
http://localhost:8080     -> Rest Server
http://localhost:8529     -> Arango DB_**

**_Everything is working on my local machine_**

Using the following Docker commands:

**_docker run -p 8529:8529 -e ARANGO_NO_AUTH=1 arangodb:3.11_**

**_docker run -it --rm absaoss/spline-admin:0.7.9 db-init arangodb://172.17.0.1/spline_**

**_docker container run -e SPLINE_DATABASE_CONNECTION_URL=arangodb://host.docker.internal/spline  -p 8080:8080 absaoss/spline-rest-server:0.7.9_**

**_docker container run -e SPLINE_CONSUMER_URL=http://localhost:8080/consumer -p 9090:8080 absaoss/spline-web-ui:0.7.5_**



**_The only link I can't access after running the docker commands is http://localhost:8080/producer_**

**Reply by cerveada** [reply]
And the spark is running where?

**Reply by wajda** [reply]
@kabir0514 It's OK that you cannot access `http://localhost:8080/producer` from the browser as it doesn't support HTTP GET method. That URL is only intended to be used by the Spline agent to send the lineage related metadata to. 
As @cerveada said, it's important that the hostname (`localhost` in your case) and the port (`TCP/8080` in your case) are accessible from the host where Spark driver is running. If you are not confident in basic networking I suggest you to ask some other colleague of you to help you with that. It should be a simple thing.

**Reply by learnerr101** [reply]
@wajda I understand the basic networking bit, so configuring it shouldn't be a problem.
As I have previoulsy mentioned that everything is configured on my local (that meant that Spark too is on my local machine) Sorry if that was a bit ambiguous in my previous response.
Since my Spark is configured locally, the hostname (localhost) and port 8080 are accessible
Also, I'm able to run that Python example in your documentation and create a lineage out of it, so I believe no networking issues are there. It is only for the Spark jobs that error crops up.
Hope that helps!

**Reply by wajda** [reply]
What does the error say, can you show it once again please?

**Reply by learnerr101** [reply]
Yeah sure!
![image](https://github.com/user-attachments/assets/44c2d909-0f65-41a9-bf92-0dbd6d77cfc1)

**Reply by wajda** [reply]
If as you said the Python example worked then the Spline server should indeed be OK.
The error message on the screenshot you posted says that the agent tried to reach the server but got `404 Not Found` response. That in turn suggests that some HTTP server is indeed listening on the Spark driver's _localhost_ on the port 8080, but it might be a different server, thus 404.
All the Spline agent is doing at that particular moment is this:
```shell
curl -I http://localhost:8080/producer/status
```
Try it and see what you've got.

**Reply by learnerr101** [reply]
Hey,

This is the output of the above curl command:

HTTP/1.1 200
Vary: Origin
Vary: Access-Control-Request-Method
Vary: Access-Control-Request-Headers
ABSA-Spline-API-Version: 1.1
ABSA-Spline-API-Version: 1
ABSA-Spline-API-LTS-Version: 1.1
ABSA-Spline-Accept-Request-Encoding: gzip
Content-Length: 0
Date: Tue, 30 Jul 2024 19:51:16 GMT

**Reply by wajda** [reply]
So, this means that it works as expected.
Something is wrong with your Spark driver network settings.
Try to use IP instead of `localhost`
